libcudacxx/docs/extended_api/synchronization_primitives/barrier/memcpy_async_tx.md

```diff
@@ -0,0 +1,87 @@
+---
+grand_parent: Extended API
+parent: Barriers
+---
+
+# `cuda::device::memcpy_async_tx`
+
+Defined in header `<cuda/barrier>`:
+
+```cuda
+template <typename T, size_t Alignment>
+inline __device__
+async_contract_fulfillment
+cuda::device::memcpy_async_tx(
+  T* dest,
+  const T* src,
+  cuda::aligned_size_t<Alignment> size,
+  cuda::barrier<cuda::thread_scope_block>& bar);
+```
+
+Copies `size` bytes from global memory `src` to shared memory `dest` and arrives
+on a shared memory barrier `bar`, updating its transaction count by `size`
+bytes.
+ 
+## Notes
+
+The behavior is undefined if any of the following conditions hold: 
+- `bar` is not in `__shared__` memory;
+- `dest` is not 16-byte aligned
+- `src` is not 16-byte aligned
+- `size` is not a multiple of 16.
+
+This function can only be used under CUDA Compute Capability 9.0 (Hopper) or
+higher.
+
+There is no feature flag to check if `cuda::device::memcpy_async_tx` is
+available.
+
+## Return Value
+
+Returns `async_contract_fulfillment::async`.
```

Comment:
I don't know what `::async` means yet (docs missing), but if the implementation is built on top of PTX, and that PTX does not guarantee asynchrony - i.e. it is conforming for that PTX to block until the copy completes - then the current implementation may block until the copy completes.

I don't recall the goals for this, but at best this will be able to say whether an async copy PTX operation was issued, and not whether this API is non-blocking / asynchronous.

Reply:
Good point. This return type is already used in the library, but oddly not documented. I have filed an issue https://github.com/NVIDIA/cccl/issues/457